local-cli/runAndroid/runAndroid.js

```diff
@@ -78,13 +78,101 @@ function tryRunAdbReverse(device) {
+  if (args.deviceId) {
+    runOnSpecificDevice(args, cmd, packageName, adbPath);
+  } else {
+    runOnAllDevices(args, cmd, packageName, adbPath);
+  }
+}
+
+function runOnSpecificDevice(args, gradlew, packageName, adbPath) {
+  let devices = adb.getDevices();
+  let selectedDevice = devices.filter(function(deviceId) {
+    return ((deviceId === args.deviceId) ? deviceId : false);
+  });
+  if (selectedDevice.length > 0) {
+    buildApk(gradlew);
+    installAndLaunchOnDevice(args, selectedDevice, packageName, adbPath);
+  } else {
+      if (devices && devices.length > 0) {
+        console.log('Could not find device with the id: "' + args.deviceId + '".');
+        console.log('Choose one of the following:');
+        console.log(adb.getDevices());
+      } else {
+        console.log('No Android devices connected.');
+      }
+  }
+}
+
+function buildApk(gradlew) {
   try {
-    adb.getDevices().map((device) => tryRunAdbReverse(device));
+    console.log(chalk.bold(
+      'Building the app...'
+    ));
+
+    // using '-x lint' in order to ignore linting errors while building the apk
+    child_process.execFileSync(gradlew, ['build', '-x', 'lint'], {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(chalk.red(
+      'Could not build the app on the device, read the error above for details.\n'
+    ));
+  }
+}
+
+function tryInstallAppOnDevice(args, device) {
+  try {
+    const pathToApk = 'app/build/outputs/apk/app-debug.apk';
+    const adbPath = getAdbPath();
+    const adbArgs = ['-s', device, 'install', pathToApk];
+    console.log(chalk.bold(
+      `Installing the app on the device (cd android && adb -s ${device} install ${pathToApk}`
+    ));
+    child_process.execFileSync(adbPath, adbArgs, {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(e.message);
+    console.log(chalk.red(
+      'Could not install the app on the device, read the error above for details.\n'
+    ));
+  }
+}
+
+function tryLaunchAppOnDevice(device, packageName, adbPath) {
+  try {
+    const adbArgs = ['-s', device, 'shell', 'am', 'start', '-n', packageName + '/.MainActivity'];
+    console.log(chalk.bold(
+      `Starting the app on ${device} (${adbPath} ${adbArgs.join(' ')})...`
+    ));
+    child_process.spawnSync(adbPath, adbArgs, {stdio: 'inherit'});
+  } catch (e) {
+    console.log(chalk.red(
+      'adb invocation failed. Do you have adb in your PATH?'
+    ));
+  }
+}
 
-    const cmd = process.platform.startsWith('win')
-      ? 'gradlew.bat'
-      : './gradlew';
+function installAndLaunchOnDevice(args, selectedDevice, packageName, adbPath) {
+  tryRunAdbReverse(selectedDevice);
+  tryInstallAppOnDevice(args, selectedDevice);
+  tryLaunchAppOnDevice(selectedDevice, packageName, adbPath);
+}
 
+function runOnAllDevices(args, cmd, packageName, adbPath){
+  try {
+    adb.getDevices().map((device) => tryRunAdbReverse(device));
```

Comment:
Should move this `tryRunAdbReverse` call into the `forEach` block below, after we've guaranteed that we can get a list of devices. Otherwise the `map` call will error and we'd never fall back to `shell am start`.